Azure Permissons (#2077)
spectrocloud/librarium: docs/docs-content/clusters/public-cloud/azure/required-permissions.md

---
sidebar_label: "Required Permissions"
title: "Required Permissions"
description: "Review the required permissions for deploying clusters on Azure"
icon: ""
hide_table_of_contents: false
tags: ["public cloud", "azure", "permissions"]
sidebar_position: 100
---

Palette requires a set of permissions to properly deploy and manage the lifecycle of clusters deployed to Azure. We recommend creating a [role assignment](https://learn.microsoft.com/en-us/azure/role-based-access-control/role-assignments) that has the service principal you want to use with Palette and the appropriate [scope level](https://learn.microsoft.com/en-us/azure/role-based-access-control/scope-overview). To ensure that Palette and VerteX can deploy and manage clusters on Azure in all use cases, use a subscription as the scope level for the role assignment.

:::warning

> We recommend you don't use the Azure build-in role, [Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#contributor), as the role assigned to the service principal you want Palette to use. Creating a custom role allows you to provide only those permissions required by Palette.

Reviewer: ```suggestion
We recommend against assigning the built-in Azure [Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#contributor) role to the service principal you want to use in Palette because its permission scope exceeds our requirements. Instead, create a custom role to provide only those permissions that Palette requires.
```

:::

We support the following use cases:

- [Static Placement](#static-placement) - Palette deploys clusters on the pre-existing network resources you specify.

- [Dynamic Placement](#dynamic-placement) - Palette creates network resources required for your cluster.

:::tip

You can use the [Validator](https://github.com/spectrocloud-labs/validator) with the [Azure plugin](https://github.com/spectrocloud-labs/validator-plugin-azure) to verify you have setup the correct permissions. The Validator will also check if the service principal used by Palette has the required permissions. The Validator Azure plugin requires the following permissions:

- Microsoft.Authorization/denyAssignments/read
- Microsoft.Authorization/roleAssignments/read
- Microsoft.Authorization/roleDefinitions/read

> You can use the Validator with the Palette CLUI. Check out the Palette CLI [Validator](../../../palette-cli/commands/validator.md) command reference page for more information.

Reviewer: ```suggestion
You can use the Validator with the Palette CLI. Check out the Palette CLI [Validator](../../../palette-cli/commands/validator.md) command reference page for more information.
```

    | Microsoft.Network/privateDnsZones/virtualNetworkLinks/write | Subscription |
    | Microsoft.Network/privateDnsZones/write | Subscription |
    | Microsoft.Network/publicIPAddresses/delete | Subscription |
    | Microsoft.Network/publicIPAddresses/join/action | Subscription |
    | Microsoft.Network/publicIPAddresses/read | Subscription |
    | Microsoft.Network/publicIPAddresses/write | Subscription |
    | Microsoft.Network/routeTables/delete | Subscription |
    | Microsoft.Network/routeTables/read | Subscription |
    | Microsoft.Network/routeTables/write | Subscription |
    | Microsoft.Network/virtualNetworks/join/action | Subscription |
    | Microsoft.Resources/subscriptions/resourceGroups/read | Subscription |
    | Microsoft.Network/virtualNetworks/read | Subscription |
    | Microsoft.Network/virtualNetworks/subnets/join/action | Subscription |
    | Microsoft.Network/virtualNetworks/subnets/read | Subscription |
    | Microsoft.Compute/galleries/images/read | Subscription |
    | Microsoft.Compute/galleries/images/versions/read| Subscription |

</TabItem>

</Tabs>

## Dynamic Placement

> Dynamic placement is the default deployment method when you are deploying Azure clusters through Palette. With dynamic placement, Palette creates the network resources required for your cluster. To support dynamic placement, ensure Palette has the following actions at the subscription level scope. 

Reviewer: ```suggestion
Dynamic placement is the default method for deploying Azure clusters through Palette. With dynamic placement, Palette dynamically creates the network resources required for your cluster. 
```

Author: Added ✔️

---
sidebar_label: "Required Permissions"
title: "Required Permissions"
description: "Review the required permissions for deploying clusters on Azure"
icon: ""
hide_table_of_contents: false
tags: ["public cloud", "azure", "permissions"]
sidebar_position: 100
---

> Palette requires a set of permissions to properly deploy and manage the lifecycle of clusters deployed on Azure. We recommend you create a custom role with all the required permissions for Palette or VerteX to deploy and manage clusters on Azure. Create a [role assignment](https://learn.microsoft.com/en-us/azure/role-based-access-control/role-assignments) that has the service principal you want to use with Palette and the appropriate [scope level](https://learn.microsoft.com/en-us/azure/role-based-access-control/scope-overview). By using the subscription as the scope level for the role assignment, you can ensure all use cases for Palette and VerteX are supported. 

Reviewer: By "in all use cases", do you mean static/dynamic placement and the Validator?

:::warning

A custom role allows you to limit the permissions to only what is required by Palette. If you don't want to create a custom role, you can assign the Azure  build-in [Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#contributor) role to the service principal used by Palette. The Contributor role has all the required permissions, but it also has additional permissions that are not required by Palette. 

:::

> The following use cases are supported:

Reviewer: ```suggestion
We support the following use cases:
```

> - [Static Placement](#static-placement) - Palette will deploy clusters on a pre-existing network resources specified by you.

Reviewer: ```suggestion
- [Static Placement](#static-placement) - Palette deploys clusters on the pre-existing network resources you specify.
```

> - [Dynamic Placement](#dynamic-placement) - Palette will create the network resources required for the cluster.

Reviewer: ```suggestion
- [Dynamic Placement](#dynamic-placement) - Palette creates network resources required for your cluster.
```

- [Validator](#validator) - Use the Validator open-source framework to validate the permissions required by Palette.

## Static Placement

> Static placement is when you have a pre-existing network resource group, VNet, subnets and security groups you want Palette to use when deploying a cluster. Static placement supports three use cases. The required actions are all the same but the scope levels for each use case are different.

Reviewer: ```suggestion
Choose static placement when you want Palette to use pre-existing network resource groups, VNets, subnets, and security groups.
```

Reviewer: Comments on the structure.

Given that we offer three options that seem to be at the same level, it makes sense to standardize the information we provide. Here's what I would do:

- Name the option (e.g., Static placement).
- Offer a broad use case (e.g., when you want Palette to use pre-existing network resource groups, VNets, subnets, and security groups).
- Introduce sub-options, if any.
- List the required actions and their scopes.

And a small note: usually, counting things in docs is not helpful unless there's an explicit need for that. When you say, "Palette supports three use cases," that doesn't add much for our readers. It can also be difficult to support in the future—when the number of use cases changes, it can be hard to find and update these sentences. Instead, we should name the use cases and treat them as a lead-in to whatever content follows.

- Microsoft.Network/networkSecurityGroups/securityRules/delete
- Microsoft.Network/networkSecurityGroups/securityRules/read
- Microsoft.Network/networkSecurityGroups/securityRules/write
- Microsoft.Network/privateDnsZones/A/delete
- Microsoft.Network/privateDnsZones/A/read
- Microsoft.Network/privateDnsZones/A/write
- Microsoft.Network/privateDnsZones/delete
- Microsoft.Network/privateDnsZones/read
- Microsoft.Network/privateDnsZones/virtualNetworkLinks/delete
- Microsoft.Network/privateDnsZones/virtualNetworkLinks/read
- Microsoft.Network/privateDnsZones/virtualNetworkLinks/write
- Microsoft.Network/privateDnsZones/write
- Microsoft.Network/publicIPAddresses/delete
- Microsoft.Network/publicIPAddresses/join/action
- Microsoft.Network/publicIPAddresses/read
- Microsoft.Network/publicIPAddresses/write
- Microsoft.Network/routeTables/delete
- Microsoft.Network/routeTables/read
- Microsoft.Network/routeTables/write
- Microsoft.Network/virtualNetworks/join/action
- Microsoft.Network/virtualNetworks/read
- Microsoft.Network/virtualNetworks/subnets/join/action
- Microsoft.Network/virtualNetworks/subnets/read
- Microsoft.Resources/subscriptions/resourceGroups/read

> </details>

Reviewer: I'm not sure we're helping readers by also providing this entire list without the Palette-specific requirements. What was your intention behind it?

> Review the table below for the required actions and the different scope levels for each use case.

Reviewer: ```suggestion
Palette supports the following use cases for static placement: Single Cluster, Multiple Cluster – Same Resource Group, and Multiple Cluster – Same Subscription. Each use case requires the same actions but with different scope levels.

Review the following tables to learn about the actions and scope levels Palette requires for each use case.
```

    | Microsoft.Network/privateDnsZones/virtualNetworkLinks/write | Subscription |
    | Microsoft.Network/privateDnsZones/write | Subscription |
    | Microsoft.Network/publicIPAddresses/delete | Subscription |
    | Microsoft.Network/publicIPAddresses/join/action | Subscription |
    | Microsoft.Network/publicIPAddresses/read | Subscription |
    | Microsoft.Network/publicIPAddresses/write | Subscription |
    | Microsoft.Network/routeTables/delete | Subscription |
    | Microsoft.Network/routeTables/read | Subscription |
    | Microsoft.Network/routeTables/write | Subscription |
    | Microsoft.Network/virtualNetworks/join/action | Subscription |
    | Microsoft.Resources/subscriptions/resourceGroups/read | Subscription |
    | Microsoft.Network/virtualNetworks/read | Subscription |
    | Microsoft.Network/virtualNetworks/subnets/join/action | Subscription |
    | Microsoft.Network/virtualNetworks/subnets/read | Subscription |
    | Microsoft.Compute/galleries/images/read | Subscription |
    | Microsoft.Compute/galleries/images/versions/read| Subscription |

</TabItem>

</Tabs>

## Dynamic Placement

> If you want to support dynamic placement use cases, ensure Palette has the following actions at the subscription level: 

Reviewer: To stay consistent with the previous section, this needs an explanation for when users should choose dynamic placement.

As you've described under static placement, that would be "when you have a pre-existing network resource group, VNet, subnets and security groups you want Palette to use when deploying a cluster". What is the use case here?

| Microsoft.Network/virtualNetworks/subnets/joinLoadBalancer/action | Subscription |
| Microsoft.Network/virtualNetworks/subnets/read       | Subscription |
| Microsoft.Network/virtualNetworks/subnets/virtualMachines/read | Subscription |
| Microsoft.Network/virtualNetworks/subnets/write      | Subscription |
| Microsoft.Network/virtualNetworks/virtualMachines/read | Subscription |
| Microsoft.Network/virtualNetworks/virtualNetworkPeerings/delete | Subscription |
| Microsoft.Network/virtualNetworks/virtualNetworkPeerings/read | Subscription |
| Microsoft.Network/virtualNetworks/virtualNetworkPeerings/write | Subscription |
| Microsoft.Network/virtualNetworks/write              | Subscription |
| Microsoft.Resources/subscriptions/resourceGroups/read | Subscription |
| Microsoft.Storage/storageAccounts/blobServices/containers/read | Subscription |
| Microsoft.Storage/storageAccounts/blobServices/containers/write | Subscription |
| Microsoft.Storage/storageAccounts/blobServices/listKeys/action | Subscription |
| Microsoft.Storage/storageAccounts/read                | Subscription |
| Microsoft.Storage/storageAccounts/write               | Subscription |

> ## Validator

Reviewer: This feels out of context here. Static and dynamic placement are types of management in Palette, and Validator is a tool. It might be the case that I don't understand the technical side of it, but purely from the structure perspective, I'd treat Validator as different from static and dynamic placement.

Reviewer 2: I wanted to jump in and add some context.

The Validator tool is something we created to help customers deploy Palette. Its purpose is to help our customers ensure that they've followed the instructions under **Static Placement** and **Dynamic Placement** correctly - that they've added the permissions required to use Palette. For Validator to work, it itself needs permissions. They're kind of "meta" permissions - permissions that allow a service principal to check permissions on other service principals.

You'e right that the act of using Validator is something completely different from the act of using Palette. We were trying to figure out the best place to tell our users which Azure permissions they need in order to use Validator.

> To use the [Validator](https://github.com/spectrocloud-labs/validator) with the [Azure plugin](https://github.com/spectrocloud-labs/validator-plugin-azure) you need the following actions to be able to ensure the roles attached to the service principal used by Validator have the required permissions:

Reviewer: When would be good cases for me to use the Validator? As you've described under static placement, that would be "when you have a pre-existing network resource group, VNet, subnets and security groups you want Palette to use when deploying a cluster". What is the use case here?

- Microsoft.Authorization/denyAssignments/read
- Microsoft.Authorization/roleAssignments/read
- Microsoft.Authorization/roleDefinitions/read

> You can use the Palette CLI to get started with the Validator. Check out the [Validator](../../../palette-cli/commands/validator.md) command reference for more information.

Reviewer: When I see this section, I assume that users looking for actions here already know what the Validator is. So, this sentence feels a little out of place here.

- [Dynamic Placement](#dynamic-placement) - Palette creates network resources required for your cluster.

:::tip

> You can use the [Validator](https://github.com/spectrocloud-labs/validator) with the [Azure plugin](https://github.com/spectrocloud-labs/validator-plugin-azure) to verify you have setup the correct permissions. The Validator will also check if the service principal used by Palette has the required permissions. The Validator Azure plugin requires the following permissions:

Reviewer 2: > to verify you have setup the correct permissions

> The Validator will also check if the service principal used by Palette has the required permissions.

I think these two statements are redundant. Aren't they both describing the validator checking whether the service principal the customer wants to use with Palette has the required permissions for using Palette?
